cub/cub/detail/choose_offset.cuh

```diff
@@ -56,6 +57,18 @@ struct ChooseOffsetT
                                          unsigned long long>::type;
 };
 
+/**
+ * common_iterator_value sets member type to the common_type of 
+ * value_type for all argument types. used to get OffsetT in 
+ * DeviceSegmentedReduce.
+ */
+template <typename... Iter>
+struct common_iterator_value {
+    using type = cuda::std::__common_type_t<cuda::std::__iter_value_type<Iter>...>;
```

Comment:
remark: I tried to figure out if we need `ChooseOffsetT` here. The idea behind `ChooseOffsetT` is to reduce the number of instantiations of CUB kernels. Segmented reduce forwards the offset iterator types all the way to the reduce agent, so using `ChooseOffsetT` wouldn't help us reduce compilation time. 

On the other hand, the underlying reduce agent was written and tested with unsigned `OffsetT` in mind. If users provide offset iterators with signed value type, they'll end up in a "not-as-well" tested area. 

For now I'm inclining towards not having `ChooseOffsetT` and adding tests to cover signed / unsigned offset iterator value types. Will live another note in the testing area.

Reply:
Sounds good. Please let me know if the tests cover what you had in mind.

Reply:
I think that keeping `OffsetT` resolution with `common_iterator_value_t` would be good because it would allow mixing iterators with different value types. For example, say that I'd like to compute the sum from `0` to `Ind1` and from `0` to `Ind2`. The user would naturally try something like
```
size_t Ind1 = <some value>;
size_t Ind2 = <some_other_value>;
auto begin_offsets = thrust::make_constant_iterator(0);
thrust::device_vector<size_t> end_offsets = {Ind1,Ind2};
```
If we constrain the two iterators to have the same `value_type`, then this code would not work, and you'd have to pass `static_cast<size_t>(0)` or `0ull` to `make_constant_iterator`.

Reply:
Oh, are you proposing simply not passing OffsetT explicitly and just doing the type inference farther down the line?

Reply:
Yeah, after reading your message again, that seems like it is what you'd like to do. You know, I think that we should probably remove OffsetT as an input parameter and just deduce it at the agent stage, like you suggested. I'd be happy to implement that change. @senior-zero Should I go ahead?

Reply:
@jecs I'm sorry for the confusion. I was trying to understand if:
```cpp
template <typename... Iter>
struct common_iterator_value
{
  using type = ::cuda::std::__common_type_t<::cuda::std::__iter_value_type<Iter>...>;
};
```
needs to be:
```cpp
template <typename... Iter>
struct common_iterator_value
{
  using type = typename ChooseOffsetT<::cuda::std::__common_type_t<::cuda::std::__iter_value_type<Iter>...>>::Type;
};
```
and concluded that it doesn't need to use `ChooseOffsetT`, but we have to test signed and unsigned offset value types in segmented reduce.